server/nuxeo-launcher/src/main/java/org/nuxeo/launcher/NuxeoLauncher.java

```diff
@@ -1695,131 +1596,235 @@ protected void setConfigProperties() throws ConfigurationException, IOException,
+            pb.redirectOutput(ProcessBuilder.Redirect.INHERIT).redirectError(ProcessBuilder.Redirect.INHERIT);
+        }
+        log.debug("Server command: {}", pb::command);
+        Process nuxeoProcess = pb.start();
+        nuxeoProcess.onExit().thenAccept(p -> {
+            if (SystemUtils.IS_OS_WINDOWS && configurationGenerator.getNuxeoHome().getPath().contains(" ")) {
+                // NXP-17679
+                log.error("The server path must not contain spaces under Windows.");
+            }
+            int exitValue = p.exitValue();
+            if (exitValue != 0) {
+                log.error("Server stopped with status: {}", exitValue);
+            }
+        });
+
+        // get pid and write it to the disk for later use
+        String pid;
+        try {
+            pid = String.valueOf(nuxeoProcess.pid());
+        } catch (UnsupportedOperationException e) {
+            log.warn("Unable to get process ID from process: {}, please report it to Nuxeo", nuxeoProcess);
+            // fallback on process manager
+            pid = processManager.findPid(processRegex).orElse(null);
+        }
+        if (pid == null) {
+            log.warn("Sent server start command but could not get process ID.");
+        } else {
+            log.warn("Server started with process ID {}.", pid);
+            File pidFile = new File(configurationGenerator.getPidDir(), "nuxeo.pid");
+            try (FileWriter writer = new FileWriter(pidFile)) {
+                writer.write(pid);
+            }
+        }
+        return nuxeoProcess;
+    }
+
+    protected void waitForEffectiveStart(Process nuxeoProcess) throws InterruptedException {
         int startMaxWait = Integer.parseInt(
-                configurationGenerator.getUserConfig().getProperty(START_MAX_WAIT_PARAM, getDefaultMaxWait()));
+                configurationGenerator.getUserConfig().getProperty(START_MAX_WAIT_PARAM, START_MAX_WAIT_DEFAULT));
+        var startTime = Instant.now();
+        var waitUntil = startTime.plusSeconds(startMaxWait);
         log.debug("Will wait for effective start during {} seconds.", startMaxWait);
         final StringBuilder startSummary = new StringBuilder();
-        final String newLine = System.getProperty("line.separator");
-        boolean isReady = false;
-        long deltaTime;
-        // Wait for status servlet ready
-        do {
+        // Wait for effective start reported from status servlet
+        boolean servletAvailable = false;
+        int n = 0;
+        while (Instant.now().isBefore(waitUntil) && nuxeoProcess.isAlive()) {
             try {
-                isReady = statusServletClient.init();
-            } catch (SocketTimeoutException e) {
-                if (!quiet) {
-                    System.out.print(".");
+                // delay will be 1s 10 times, then 2s 10 times... until reaching maximum of 1min
+                Thread.sleep(Math.min((n % 10 + 1) * 1000, 60_000));
```

Comment:
this is weird I don't see where n is incremented and if it is you will never reach 1min (retries will be spaced by 1s, 2s, ... 9s, 10s, then again 1s, ...)

Reply:
Good catch 👍. I tried different approaches on that without finding a way to detect fast startup (~22s). Do you have a clue to rework this waiting aspect?

Reply:
Changed by:
```java
Thread.sleep(Math.min((++n / 10 + 1) * 1000, 60_000));
```

Reply:
maybe just trying every second is good enough then